File: client/src/program/instructions.ts
import { PublicKey } from "@solana/web3.js";

export class TokenLockInstruction {
  static Initialize({
    interest_mint,
    collateral_mint,
    base_interest_rate,
    price_factor,
    min_commission_rate,
    max_commission_rate,
    min_deposit_amount,
    max_deposit_amount,
    deposit_periods,
  }: {
    interest_mint: PublicKey;
    collateral_mint: PublicKey;
    base_interest_rate: number;
    price_factor: number;
    min_commission_rate: number;
    max_commission_rate: number;
    min_deposit_amount: number;
    max_deposit_amount: number;
    deposit_periods: number[];
  }): TokenLockInstruction {
    return new TokenLockInstruction("Initialize", {
      interest_mint,
      collateral_mint,
      base_interest_rate,
      price_factor,
      min_commission_rate,
      max_commission_rate,
      min_deposit_amount,
      max_deposit_amount,
      deposit_periods,
    });
  }

  static AdminUpdateConfig({
    param,
    base_interest_rate,
    price_factor,
    min_commission_rate,
    max_commission_rate,
    min_deposit_amount,
    max_deposit_amount,
    deposit_periods,
  }: {
    param: number;
    base_interest_rate?: number;
    price_factor?: number;
    min_commission_rate?: number;
    max_commission_rate?: number;
    min_deposit_amount?: number;
    max_deposit_amount?: number;
    deposit_periods?: number[];
  }): TokenLockInstruction {
    return new TokenLockInstruction("AdminUpdateConfig", {
      param,
      base_interest_rate,
      price_factor,
      min_commission_rate,
      max_commission_rate,
      min_deposit_amount,
      max_deposit_amount,
      deposit_periods,
    });
  }

  static AdminWithdrawCollateralForInvestment(): TokenLockInstruction {
    return new TokenLockInstruction("AdminWithdrawCollateralForInvestment", {});
  }

  static AdminPrepareWithdrawal(): TokenLockInstruction {
    return new TokenLockInstruction("AdminPrepareWithdrawal", {});
  }

  static AdminDepositInterest({
    amount,
  }: {
    amount: number;
  }): TokenLockInstruction {
    return new TokenLockInstruction("AdminDepositInterest", { amount });
  }

  static AdminWithdrawInterest({
    amount,
  }: {
    amount: number;
  }): TokenLockInstruction {
    return new TokenLockInstruction("AdminWithdrawInterest", { amount });
  }

  static DepositCollateral({
    amount,
    deposit_period,
    commission_rate,
  }: {
    amount: number;
    deposit_period: number;
    commission_rate: number;
  }): TokenLockInstruction {
    return new TokenLockInstruction("DepositCollateral", {
      amount,
      deposit_period,
      commission_rate,
    });
  }

  static RequestWithdrawalEarly(): TokenLockInstruction {
    return new TokenLockInstruction("RequestWithdrawalEarly", {});
  }

  static RequestWithdrawal(): TokenLockInstruction {
    return new TokenLockInstruction("RequestWithdrawal", {});
  }

  static WithdrawCollateral(): TokenLockInstruction {
    return new TokenLockInstruction("WithdrawCollateral", {});
  }

  constructor(
    public readonly type: string,
    public readonly params: Record<string, unknown>
  ) {}

  private numberToLEBytes(num: number, decimals: number = 0): Uint8Array {
    const intValue = Math.round(num * Math.pow(10, decimals));
    return new Uint8Array(new BigUint64Array([BigInt(intValue)]).buffer);
  }

  pack(): Buffer {
    const buffer: number[] = [];

    switch (this.type) {
      case "Initialize": {
        const {
          interest_mint,
          collateral_mint,
          base_interest_rate,
          price_factor,
          min_commission_rate,
          max_commission_rate,
          min_deposit_amount,
          max_deposit_amount,
          deposit_periods,
        } = this.params as {
          interest_mint: PublicKey;
          collateral_mint: PublicKey;
          base_interest_rate: number;
          price_factor: number;
          min_commission_rate: number;
          max_commission_rate: number;
          min_deposit_amount: number;
          max_deposit_amount: number;
          deposit_periods: number[];
        };

        buffer.push(0);

        const interestMintBytes = interest_mint.toBytes();
        for (let i = 0; i < interestMintBytes.length; i++) {
          buffer.push(interestMintBytes[i]);
        }

        const collateralMintBytes = collateral_mint.toBytes();
        for (let i = 0; i < collateralMintBytes.length; i++) {
          buffer.push(collateralMintBytes[i]);
        }

        const baseInterestBytes = this.numberToLEBytes(base_interest_rate);
        const priceFactorBytes = this.numberToLEBytes(price_factor);
        const minCommissionBytes = this.numberToLEBytes(min_commission_rate);
        const maxCommissionBytes = this.numberToLEBytes(max_commission_rate);
        const minDepositBytes = this.numberToLEBytes(min_deposit_amount);
        const maxDepositBytes = this.numberToLEBytes(max_deposit_amount);

        for (let i = 0; i < baseInterestBytes.length; i++)
          buffer.push(baseInterestBytes[i]);
        for (let i = 0; i < priceFactorBytes.length; i++)
          buffer.push(priceFactorBytes[i]);
        for (let i = 0; i < minCommissionBytes.length; i++)
          buffer.push(minCommissionBytes[i]);
        for (let i = 0; i < maxCommissionBytes.length; i++)
          buffer.push(maxCommissionBytes[i]);
        for (let i = 0; i < minDepositBytes.length; i++)
          buffer.push(minDepositBytes[i]);
        for (let i = 0; i < maxDepositBytes.length; i++)
          buffer.push(maxDepositBytes[i]);

        for (const period of deposit_periods) {
          const periodBytes = this.numberToLEBytes(period);
          for (let i = 0; i < periodBytes.length; i++)
            buffer.push(periodBytes[i]);
        }
        break;
      }
      case "AdminUpdateConfig": {
        const {
          param,
          base_interest_rate,
          price_factor,
          min_commission_rate,
          max_commission_rate,
          min_deposit_amount,
          max_deposit_amount,
          deposit_periods,
        } = this.params as {
          param: number;
          base_interest_rate?: number;
          price_factor?: number;
          min_commission_rate?: number;
          max_commission_rate?: number;
          min_deposit_amount?: number;
          max_deposit_amount?: number;
          deposit_periods?: number[];
        };

        buffer.push(1);
        buffer.push(param);

        if (base_interest_rate !== undefined) {
          buffer.push(1);
          const rateBytes = this.numberToLEBytes(base_interest_rate);
          for (let i = 0; i < rateBytes.length; i++) buffer.push(rateBytes[i]);
        } else {
          buffer.push(0);
        }

        if (price_factor !== undefined) {
          buffer.push(1);
          const factorBytes = this.numberToLEBytes(price_factor);
          for (let i = 0; i < factorBytes.length; i++)
            buffer.push(factorBytes[i]);
        } else {
          buffer.push(0);
        }

        if (min_commission_rate !== undefined) {
          buffer.push(1);
          const rateBytes = this.numberToLEBytes(min_commission_rate);
          for (let i = 0; i < rateBytes.length; i++) buffer.push(rateBytes[i]);
        } else {
          buffer.push(0);
        }

        if (max_commission_rate !== undefined) {
          buffer.push(1);
          const rateBytes = this.numberToLEBytes(max_commission_rate);
          for (let i = 0; i < rateBytes.length; i++) buffer.push(rateBytes[i]);
        } else {
          buffer.push(0);
        }

        if (min_deposit_amount !== undefined) {
          buffer.push(1);
          const amountBytes = this.numberToLEBytes(min_deposit_amount);
          for (let i = 0; i < amountBytes.length; i++)
            buffer.push(amountBytes[i]);
        } else {
          buffer.push(0);
        }

        if (max_deposit_amount !== undefined) {
          buffer.push(1);
          const amountBytes = this.numberToLEBytes(max_deposit_amount);
          for (let i = 0; i < amountBytes.length; i++)
            buffer.push(amountBytes[i]);
        } else {
          buffer.push(0);
        }

        if (deposit_periods !== undefined) {
          buffer.push(1);
          for (const period of deposit_periods) {
            const periodBytes = this.numberToLEBytes(period);
            for (let i = 0; i < periodBytes.length; i++)
              buffer.push(periodBytes[i]);
          }
        } else {
          buffer.push(0);
        }

        break;
      }
      case "AdminWithdrawCollateralForInvestment": {
        buffer.push(2);
        break;
      }
      case "AdminPrepareWithdrawal": {
        buffer.push(3);
        break;
      }
      case "AdminDepositInterest": {
        const { amount } = this.params as { amount: number };
        buffer.push(4);
        const amountBytes = this.numberToLEBytes(amount, 6);
        for (let i = 0; i < amountBytes.length; i++)
          buffer.push(amountBytes[i]);
        break;
      }
      case "AdminWithdrawInterest": {
        const { amount } = this.params as { amount: number };
        buffer.push(5);
        const amountBytes = this.numberToLEBytes(amount, 6);
        for (let i = 0; i < amountBytes.length; i++)
          buffer.push(amountBytes[i]);
        break;
      }
      case "DepositCollateral": {
        const { amount, deposit_period, commission_rate } = this.params as {
          amount: number;
          deposit_period: number;
          commission_rate: number;
        };
        buffer.push(6);

        const amountBytes = this.numberToLEBytes(amount, 8);
        for (let i = 0; i < amountBytes.length; i++)
          buffer.push(amountBytes[i]);

        const periodBytes = this.numberToLEBytes(deposit_period);
        for (let i = 0; i < periodBytes.length; i++)
          buffer.push(periodBytes[i]);

        const commissionBytes = this.numberToLEBytes(commission_rate, 1);
        for (let i = 0; i < commissionBytes.length; i++)
          buffer.push(commissionBytes[i]);
        break;
      }
      case "RequestWithdrawalEarly": {
        buffer.push(7);
        break;
      }
      case "RequestWithdrawal": {
        buffer.push(8);
        break;
      }
      case "WithdrawCollateral": {
        buffer.push(9);
        break;
      }
    }

    return Buffer.from(buffer);
  }
}

File: program/src/instructions.rs
use borsh::{BorshDeserialize, BorshSerialize};
use solana_program::{program_error::ProgramError, pubkey::Pubkey};

#[derive(BorshSerialize, BorshDeserialize, Debug, Clone)]
pub enum TokenLockInstruction {
    /// Initialize the pool with configuration and create necessary PDAs
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[writable]` Config PDA account
    /// 2. `[writable]` Authority PDA account
    /// 3. `[writable]` Interest pool ATA account
    /// 4. `[writable]` Collateral pool ATA account
    /// 5. `[writable]` Withdrawal pool account
    /// 6. `[]` Interest mint account
    /// 7. `[]` Collateral mint account
    /// 8. `[]` System program
    /// 9. `[]` Token program
    /// 10. `[]` Associated Token Account program
    /// 11. `[]` Rent sysvar
    Initialize {
        interest_mint: Pubkey,
        collateral_mint: Pubkey,
        base_interest_rate: u64,
        price_factor: u64,
        min_commission_rate: u64,
        max_commission_rate: u64,
        min_deposit_amount: u64,
        max_deposit_amount: u64,
        deposit_periods: Vec<u64>,
    },

    /// Update pool configuration parameters
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[writable]` Config PDA account
    AdminUpdateConfig {
        param: u8,
        base_interest_rate: Option<u64>,
        price_factor: Option<u64>,
        min_commission_rate: Option<u64>,
        max_commission_rate: Option<u64>,
        min_deposit_amount: Option<u64>,
        max_deposit_amount: Option<u64>,
        deposit_periods: Option<Vec<u64>>,
    },

    /// Admin withdraws collateral for investment
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` Admin's collateral token account
    /// 4. `[writable]` Pool's collateral token account
    /// 5. `[]` System program
    /// 6. `[]` Token program
    /// 7. `[]` Associated Token Account program
    AdminWithdrawCollateralForInvestment,

    /// Admin prepares withdrawal by depositing collateral
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[]` Config PDA account
    /// 2. `[writable]` Admin's collateral token account
    /// 3. `[writable]` Withdrawal pool account
    /// 4. `[]` User account
    /// 5. `[writable]` User deposit account
    /// 6. `[]` Token program
    AdminPrepareWithdrawal,

    /// Admin deposits interest tokens to the pool
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` Admin's interest token account
    /// 4. `[writable]` Pool's interest token account
    /// 5. `[]` System program
    /// 6. `[]` Token program
    /// 7. `[]` Associated Token Account program
    AdminDepositInterest { amount: u64 },

    /// Admin withdraws interest tokens from the pool
    ///
    /// Accounts expected:
    /// 0. `[signer]` Admin account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` Admin's interest token account
    /// 4. `[writable]` Pool's interest token account
    /// 5. `[]` System program
    /// 6. `[]` Token program
    /// 7. `[]` Associated Token Account program
    AdminWithdrawInterest { amount: u64 },

    /// Deposit collateral tokens into the pool
    ///
    /// Accounts expected:
    /// 0. `[signer]` User account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` User's collateral token account
    /// 4. `[writable]` User's deposit account
    /// 5. `[writable]` Pool's collateral token account
    /// 6. `[writable]` User's interest token account
    /// 7. `[writable]` Pool's interest token account
    /// 8. `[]` System program
    /// 9. `[]` Token program
    DepositCollateral {
        amount: u64,
        deposit_period: u64,
        commission_rate: u64,
    },

    /// Request early withdrawal of collateral (before unlock time)
    ///
    /// Accounts expected:
    /// 0. `[signer]` User account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` User's deposit account
    /// 4. `[writable]` User's interest token account
    /// 5. `[writable]` Pool's interest token account
    /// 6. `[]` Token program
    RequestWithdrawalEarly,

    /// Request withdrawal of collateral (after unlock time)
    ///
    /// Accounts expected:
    /// 0. `[signer]` User account
    /// 1. `[writable]` User's deposit account
    RequestWithdrawal,

    /// Withdraw collateral after admin preparation
    ///
    /// Accounts expected:
    /// 0. `[signer]` User account
    /// 1. `[]` Config PDA account
    /// 2. `[]` Authority PDA account
    /// 3. `[writable]` User's deposit account
    /// 4. `[writable]` User's collateral token account
    /// 5. `[writable]` Withdrawal pool account
    /// 6. `[]` Token program
    WithdrawCollateral,
}

impl TokenLockInstruction {
    pub fn unpack(input: &[u8]) -> Result<Self, ProgramError> {
        let instruction = Self::try_from_slice(input)?;
        Ok(instruction)
    }
    pub fn pack(&self) -> Result<Vec<u8>, ProgramError> {
        let mut buffer = vec![];
        match self {
            Self::Initialize {
                interest_mint,
                collateral_mint,
                base_interest_rate,
                price_factor,
                min_commission_rate,
                max_commission_rate,
                min_deposit_amount,
                max_deposit_amount,
                deposit_periods,
            } => {
                buffer.push(0);
                buffer.extend_from_slice(&interest_mint.to_bytes());
                buffer.extend_from_slice(&collateral_mint.to_bytes());
                buffer.extend_from_slice(&base_interest_rate.to_le_bytes());
                buffer.extend_from_slice(&price_factor.to_le_bytes());
                buffer.extend_from_slice(&min_commission_rate.to_le_bytes());
                buffer.extend_from_slice(&max_commission_rate.to_le_bytes());
                buffer.extend_from_slice(&min_deposit_amount.to_le_bytes());
                buffer.extend_from_slice(&max_deposit_amount.to_le_bytes());
                buffer.extend_from_slice(
                    &deposit_periods
                        .iter()
                        .flat_map(|&x| x.to_le_bytes())
                        .collect::<Vec<u8>>(),
                );
            }
            Self::AdminUpdateConfig {
                param,
                base_interest_rate,
                price_factor,
                min_commission_rate,
                max_commission_rate,
                min_deposit_amount,
                max_deposit_amount,
                deposit_periods,
            } => {
                buffer.push(1);
                buffer.push(*param);
                if let Some(rate) = base_interest_rate {
                    buffer.push(1);
                    buffer.extend_from_slice(&rate.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(factor) = price_factor {
                    buffer.push(1);
                    buffer.extend_from_slice(&factor.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(min_commission_rate) = min_commission_rate {
                    buffer.push(1);
                    buffer.extend_from_slice(&min_commission_rate.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(max_commission_rate) = max_commission_rate {
                    buffer.push(1);
                    buffer.extend_from_slice(&max_commission_rate.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(min_deposit_amount) = min_deposit_amount {
                    buffer.push(1);
                    buffer.extend_from_slice(&min_deposit_amount.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(max_deposit_amount) = max_deposit_amount {
                    buffer.push(1);
                    buffer.extend_from_slice(&max_deposit_amount.to_le_bytes());
                } else {
                    buffer.push(0);
                }
                if let Some(deposit_periods) = deposit_periods {
                    buffer.push(1);
                    buffer.extend_from_slice(
                        &deposit_periods
                            .iter()
                            .flat_map(|&x| x.to_le_bytes())
                            .collect::<Vec<u8>>(),
                    );
                } else {
                    buffer.push(0);
                }
            }
            Self::AdminWithdrawCollateralForInvestment => {
                buffer.push(2);
            }
            Self::AdminPrepareWithdrawal => {
                buffer.push(3);
            }
            Self::AdminDepositInterest { amount } => {
                buffer.push(4);
                buffer.extend_from_slice(&amount.to_le_bytes());
            }
            Self::AdminWithdrawInterest { amount } => {
                buffer.push(5);
                buffer.extend_from_slice(&amount.to_le_bytes());
            }
            Self::DepositCollateral {
                amount,
                deposit_period,
                commission_rate: comminsion_rate,
            } => {
                buffer.push(6);
                buffer.extend_from_slice(&amount.to_le_bytes());
                buffer.extend_from_slice(&deposit_period.to_le_bytes());
                buffer.extend_from_slice(&comminsion_rate.to_le_bytes());
            }
            Self::RequestWithdrawalEarly => {
                buffer.push(7);
            }
            Self::RequestWithdrawal => {
                buffer.push(8);
            }
            Self::WithdrawCollateral => {
                buffer.push(9);
            }
        }
        Ok(buffer)
    }
}
